public/app/core/services/Sonifier.ts

```diff
@@ -0,0 +1,299 @@
+export type Note = 'A' | 'Bb' | 'B' | 'C' | 'Db' | 'D' | 'Eb' | 'E' | 'F' | 'Gb' | 'G' | 'Ab';
+
+const A4_FREQUENCY = 440;
+
+const SemitoneMapping: { [note: string]: number } = {
+  A: 0,
+  Bb: 1,
+  B: 2,
+  C: 3,
+  Dd: 4,
+  D: 5,
+  Eb: 6,
+  E: 7,
+  F: 8,
+  Gb: 9,
+  G: 10,
+  Ab: 11,
+};
+
+type Scale = number[];
+type ScaleType = 'major' | 'minor' | 'chromatic';
+type ProcessFn = (() => void) | undefined;
+type SampleFrame<T> = {
+  data: T;
+  index: number;
+};
+type PlaySeriesOptions = {
+  min?: number;
+  max?: number;
+  onPointProcess?: (pointIndex: number) => void;
+};
+
+const Scales: { [name in 'major' | 'minor' | 'chromatic']: Scale } = {
+  chromatic: [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],
+  major: [2, 2, 1, 2, 2, 2, 1, 2],
+  minor: [2, 1, 2, 2, 1, 3, 1, 2],
+};
+
+export type Tuple = [ts: number, val: number];
+type SamplingMethod = 'random' | 'systematic';
+
+type SonifierOptions = {
+  baseFrequency?: number;
+  scales?: number;
+  maxSamples?: number;
+  samplingMethod?: SamplingMethod;
+  instrument?: OscillatorType;
+  scale?: ScaleType;
+  volume?: number;
+  fixedDeltaTime: boolean;
+};
+
+export class Sonifier {
+  isPlaying: boolean;
+  baseFrequency: number;
+  scales: number;
+  maxSamples: number;
+  samplingMethod: SamplingMethod;
+  instrument: OscillatorType;
+  scale: ScaleType;
+
+  private _audioContext: AudioContext;
+  private _volume: number;
+  private _paused: boolean;
+  private _fixedDt: boolean;
+  private _playQueue: Array<{ data: Tuple; duration: number; onProcess: ProcessFn }>;
+
+  constructor(options?: SonifierOptions) {
+    this.isPlaying = false;
+    this.baseFrequency = options?.baseFrequency || A4_FREQUENCY;
+    this.scales = options?.scales || 3;
+    this.maxSamples = options?.maxSamples || 50;
+    this.samplingMethod = options?.samplingMethod || 'systematic';
+    this.instrument = options?.instrument || 'square';
+    this.scale = options?.scale || 'chromatic';
+
+    this._volume = options?.volume || 0.03;
+    this._fixedDt = options?.fixedDeltaTime || false;
+    this._playQueue = [];
+    this._paused = false;
+    this._audioContext = new AudioContext();
```

Comment:
@kpelelis sometimes this audio context is stopped, because it needs a prior interaction of the user on the page (to prevent loud noise that's unwanted). Perhaps we can make instantiation lazy.

Reply:
I am wondering how is this affected as the user needs to interact with the page to start the sonification (i.e. click or press a key)